File: Problems/LongestValidParentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    let stack = [] , aux = Array(s.length+1).fill(0), result=0;
    for(let i=0; i<s.length;i++){
        if(s.charAt(i) == '(') stack.push(i);
        else{
            if(stack.length != 0){
                let lastValue = stack.pop()
                aux[i+1] += (2 + aux[i] + aux[lastValue])
                result = Math.max(result,aux[i+1])
            }
        }        
    }
    return result;
};

File: Problems/LongestValidParentheses.py
from collections import deque

openParenthesis = '('

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = deque([0])
        push, pop = stack.append, stack.pop
        
        result = 0
        
        for i in s:
            if i  == openParenthesis:
                push(0)
            elif len(stack) > 1:
                paraSize = pop() + 2
                stack[-1] += paraSize
            else:
                result = max(result, stack[0])
                stack[0] = 0
                
        return max(result, max(stack))
